File: apps/viewer/lib/broker_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:supabase_flutter/supabase_flutter.dart';

import 'config.dart';

/// Thin client for the Church-login auth broker (backend/auth_broker). The browser can't
/// call the Church's Okta directly (CORS); the broker does it server-side and hands back a
/// Supabase session OTP the app verifies. Empty [brokerUrl] => Church login is unavailable.
class BrokerException implements Exception {
  final String message;
  BrokerException(this.message);
  @override
  String toString() => message;
}

/// One MFA factor offered by Okta (e.g. "Text message to •••1234").
class BrokerFactor {
  final String id;
  final String label;
  final String method;
  BrokerFactor(this.id, this.label, this.method);
  factory BrokerFactor.fromJson(Map<String, dynamic> j) =>
      BrokerFactor(j['id'] as String, (j['label'] ?? j['method'] ?? 'Code') as String,
          (j['method'] ?? '') as String);
}

/// Result of a step: either a verifiable Supabase session, or an MFA challenge to continue.
class BrokerResult {
  final String? email; // present when a session was minted
  final String? otp; // Supabase OTP to verifyOtp(email, otp)
  final String? loginId; // present when MFA is required
  final List<BrokerFactor> factors;
  final String? name;
  /// N2: whether the signed-in calling has covenant-path access (from the broker enroll step).
  /// null = unknown (not enrolled / errored) → don't block; false = no access → block at login.
  final bool? authorized;
  BrokerResult(
      {this.email, this.otp, this.loginId, this.factors = const [], this.name, this.authorized});

  bool get mfaRequired => loginId != null && otp == null;
}

/// Credential info returned from /auth/enrollment-status.
class CredentialInfo {
  final String state; // "none" | "active" | "revoked"
  final bool complete;
  final String? principalName;
  final bool isProvider;
  final String? enrolledAt;
  const CredentialInfo({required this.state, this.complete = false,
      this.principalName, this.isProvider = false, this.enrolledAt});
  factory CredentialInfo.fromJson(Map<String, dynamic> j) => CredentialInfo(
      state: (j['state'] as String?) ?? 'none',
      complete: j['complete'] as bool? ?? false,
      principalName: j['principal_name'] as String?,
      isProvider: j['is_provider'] as bool? ?? false,
      enrolledAt: j['enrolled_at'] as String?);
  bool get isActive => state == 'active';
  bool get isRevoked => state == 'revoked';
  bool get isNone => state == 'none';
}

/// Response from /auth/enrollment-status.
class EnrollmentStatus {
  final String? stakeName;
  final String? stakeId;
  final String? lastSyncedAt;
  final int memberCount;
  final bool hasData;
  final bool noRole; // true if the user has no user_roles row yet
  final CredentialInfo credential;
  const EnrollmentStatus({this.stakeName, this.stakeId, this.lastSyncedAt,
      this.memberCount = 0, this.hasData = false, this.noRole = false,
      required this.credential});
  factory EnrollmentStatus.fromJson(Map<String, dynamic> j) {
    final isNoRole = j['status'] == 'no_role';
    return EnrollmentStatus(
        stakeName: j['stake_name'] as String?,
        stakeId: j['stake_id'] as String?,
        lastSyncedAt: j['last_synced_at'] as String?,
        memberCount: (j['member_count'] as num?)?.toInt() ?? 0,
        hasData: j['has_data'] as bool? ?? false,
        noRole: isNoRole,
        credential: CredentialInfo.fromJson(
            (j['credential'] as Map?)?.cast<String, dynamic>() ?? const {}));
  }
}

class BrokerClient {
  bool get available => brokerUrl.isNotEmpty;

  /// Called when a network attempt fails and we're about to retry — lets the UI show a
  /// "waking up the sign-in service…" message. Set once by the login page.
  void Function(String message)? onStatus;

  /// N5: wake the free-tier broker early — fire a cheap /health ping when the login screen appears,
  /// so it spins up WHILE the user types their credentials, hiding the ~30-60s cold start that
  /// otherwise lands on the first login request. Fire-and-forget; errors ignored.
  void warmUp() {
    if (!available) return;
    http
        .get(Uri.parse('$brokerUrl/health'))
        .timeout(const Duration(seconds: 30))
        .then((_) {}, onError: (_) {});
  }

  // Free hosting (Render) sleeps when idle; the first request after a sleep hits a holding
  // page with no CORS header (browser reports "Failed to fetch"). Retry across ~60s so a
  // cold start resolves itself instead of erroring out. Delays sum to ~63s.
  static const _retryDelays = [Duration(seconds: 3), Duration(seconds: 6),
      Duration(seconds: 9), Duration(seconds: 12), Duration(seconds: 15), Duration(seconds: 18)];

  /// POST with cold-start retry; returns the decoded JSON (throws BrokerException on error).
  Future<Map<String, dynamic>> _postJson(String path, Map<String, dynamic> body) async {
    if (!available) throw BrokerException('Church login is not configured (BROKER_URL).');
    http.Response? resp;
    Object? lastErr;
    for (var attempt = 0; attempt <= _retryDelays.length; attempt++) {
      try {
        resp = await http
            .post(Uri.parse('$brokerUrl$path'),
                headers: const {'Content-Type': 'application/json'}, body: jsonEncode(body))
            .timeout(const Duration(seconds: 30));
        break; // got an HTTP response (success or error) — stop retrying
      } catch (e) {
        // Network failure (cold start / transient). Wait and retry within budget.
        lastErr = e;
        if (attempt < _retryDelays.length) {
          onStatus?.call('Waking up the sign-in service… this can take up to a minute on first use.');
          await Future.delayed(_retryDelays[attempt]);
        }
      }
    }
    if (resp == null) {
      throw BrokerException(
          'Could not reach the sign-in service after several tries. It may be starting up — '
          'please try again in a minute. ($lastErr)');
    }
    Map<String, dynamic> data;
    try {
      data = jsonDecode(resp.body) as Map<String, dynamic>;
    } catch (_) {
      throw BrokerException('Sign-in service error (${resp.statusCode}).');
    }
    if (resp.statusCode >= 400) {
      throw BrokerException((data['detail'] ?? 'Sign-in failed (${resp.statusCode}).').toString());
    }
    return data;
  }

  Future<BrokerResult> _post(String path, Map<String, dynamic> body) async {
    final data = await _postJson(path, body);
    if (data['status'] == 'mfa_required') {
      return BrokerResult(
        loginId: data['login_id'] as String,
        factors: ((data['factors'] as List?) ?? const [])
            .map((f) => BrokerFactor.fromJson(f as Map<String, dynamic>))
            .toList(),
      );
    }
    final session = (data['session'] as Map?)?.cast<String, dynamic>() ?? const {};
    final enroll = (data['enroll'] as Map?)?.cast<String, dynamic>();
    return BrokerResult(
      email: session['email'] as String?,
      otp: session['otp'] as String?,
      name: data['identity_name'] as String?,
      // N2: only a present, explicit `false` blocks; absent/errored enroll → null (don't block).
      authorized: enroll != null && enroll.containsKey('authorized')
          ? enroll['authorized'] as bool?
          : null,
    );
  }

  Future<BrokerResult> password(String username, String password, {bool enroll = false}) =>
      _post('/auth/password', {'username': username, 'password': password, 'enroll': enroll});

  Future<void> selectFactor(String loginId, String factorId) =>
      _post('/auth/mfa/select', {'login_id': loginId, 'factor_id': factorId});

  Future<BrokerResult> verifyMfa(String loginId, String code, {bool enroll = false}) =>
      _post('/auth/mfa/verify', {'login_id': loginId, 'code': code, 'enroll': enroll});

  /// Email-OTP relay (for networks that can't reach Supabase directly): the broker asks Supabase
  /// to email a one-time code. The user enters it; emailVerify returns session tokens to adopt.
  Future<void> emailStart(String email) async {
    await _postJson('/auth/email/start', {'email': email});
  }

  /// Verify the emailed code via the broker → {access_token, refresh_token} for setSession.
  Future<Map<String, dynamic>> emailVerify(String email, String code) =>
      _postJson('/auth/email/verify', {'email': email, 'code': code});

  Future<EnrollmentStatus> enrollmentStatus() async {
    if (!available) throw BrokerException('Broker not configured.');
    final token = Supabase.instance.client.auth.currentSession?.accessToken ?? '';
    if (token.isEmpty) throw BrokerException('Not signed in.');
    final resp = await http
        .get(Uri.parse('$brokerUrl/auth/enrollment-status'),
            headers: {'Authorization': 'Bearer $token'})
        .timeout(const Duration(seconds: 20));
    Map<String, dynamic> data;
    try {
      data = jsonDecode(resp.body) as Map<String, dynamic>;
    } catch (_) {
      throw BrokerException('Enrollment status error (${resp.statusCode}).');
    }
    if (resp.statusCode >= 400) {
      throw BrokerException((data['detail'] ?? 'Enrollment status failed.').toString());
    }
    return EnrollmentStatus.fromJson(data);
  }

  Future<void> revoke(String stakeId) async {
    await _authed('POST', '/auth/revoke', body: {'stake_id': stakeId});
  }

  /// Provider triggers a sync for their own stake. Returns {coverage_complete, last_synced_at}.
  Future<Map<String, dynamic>> syncNow() => _authed('POST', '/auth/sync-now');

  /// In-app daily-sync schedule (#schedule): read {eligible, hour_et, paused}; write hour + pause.
  Future<Map<String, dynamic>> getSchedule() => _authed('GET', '/auth/schedule');
  Future<Map<String, dynamic>> setSchedule(int hourEt, bool paused) =>
      _authed('POST', '/auth/schedule', body: {'hour_et': hourEt, 'paused': paused});

  /// M7 per-stake Google Drive: status, start the OAuth (returns {url} to open), disconnect.
  Future<Map<String, dynamic>> googleDriveStatus() => _authed('GET', '/auth/google/status');
  Future<Map<String, dynamic>> googleDriveStart() => _authed('POST', '/auth/google/start');
  Future<Map<String, dynamic>> googleDriveDisconnect() => _authed('POST', '/auth/google/disconnect');

  /// Support form (#74): email the owner a message from the signed-in user.
  Future<void> contact(String subject, String message) async {
    await _authed('POST', '/contact', body: {'subject': subject, 'message': message});
  }

  /// Ad-hoc leader report (#73): structured convert-integration status for the caller's scope.
  Future<Map<String, dynamic>> report() async {
    return await _authed('GET', '/report');
  }

  /// Email the caller's scope report (default: to themselves; or a chosen recipient).
  Future<Map<String, dynamic>> emailReport({String? toEmail}) async {
    return await _authed('POST', '/report/email',
        body: {if (toEmail != null && toEmail.isNotEmpty) 'to_email': toEmail});
  }

  /// Shared authed request carrying the signed-in user's Supabase token. Returns the decoded
  /// JSON body (empty map if none). Throws BrokerException on transport/HTTP error.
  Future<Map<String, dynamic>> _authed(String method, String path,
      {Map<String, dynamic>? body}) async {
    if (!available) throw BrokerException('Broker not configured.');
    final token = Supabase.instance.client.auth.currentSession?.accessToken ?? '';
    if (token.isEmpty) throw BrokerException('Not signed in.');
    final uri = Uri.parse('$brokerUrl$path');
    final headers = {'Authorization': 'Bearer $token', 'Content-Type': 'application/json'};
    final resp = await (method == 'GET'
            ? http.get(uri, headers: headers)
            : http.post(uri, headers: headers, body: jsonEncode(body ?? const {})))
        .timeout(const Duration(seconds: 30));
    Map<String, dynamic> data = const {};
    if (resp.body.isNotEmpty) {
      try {
        data = jsonDecode(resp.body) as Map<String, dynamic>;
      } catch (_) {
        if (resp.statusCode >= 400) throw BrokerException('Request failed (${resp.statusCode}).');
      }
    }
    if (resp.statusCode >= 400) {
      throw BrokerException((data['detail'] ?? 'Request failed (${resp.statusCode}).').toString());
    }
    return data;
  }
}

File: native/android/app/src/main/java/org/membercovenantpath/viewer/data/BrokerClient.kt
package org.membercovenantpath.viewer.data

import io.ktor.client.statement.HttpResponse
import kotlinx.coroutines.delay
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/** Mirrors broker_client.dart's exceptions/result types so the UI logic ports 1:1. */
class BrokerException(message: String) : Exception(message)

/** One MFA factor offered by Okta (e.g. "Text message to •••1234"). */
data class BrokerFactor(val id: String, val label: String, val method: String) {
    companion object {
        fun from(j: JsonObject) = BrokerFactor(
            id = j.str("id") ?: "",
            label = j.str("label") ?: j.str("method") ?: "Code",
            method = j.str("method") ?: "",
        )
    }
}

/** Result of a broker step: a verifiable Supabase session, or an MFA challenge to continue. */
data class BrokerResult(
    val email: String? = null,
    val otp: String? = null,
    val loginId: String? = null,
    val factors: List<BrokerFactor> = emptyList(),
    val name: String? = null,
    // N2: covenant-path access from the broker enroll step. null = unknown (don't block);
    // false = no access -> block at login.
    val authorized: Boolean? = null,
) {
    val mfaRequired: Boolean get() = loginId != null && otp == null
}

/** Credential state from /auth/enrollment-status. */
data class CredentialInfo(
    val state: String = "none",
    val complete: Boolean = false,
    val principalName: String? = null,
    val isProvider: Boolean = false,
    val enrolledAt: String? = null,
) {
    val isActive get() = state == "active"
    val isRevoked get() = state == "revoked"
    val isNone get() = state == "none"

    companion object {
        fun from(j: JsonObject?) = CredentialInfo(
            state = j?.str("state") ?: "none",
            complete = j?.bool("complete") ?: false,
            principalName = j?.str("principal_name"),
            isProvider = j?.bool("is_provider") ?: false,
            enrolledAt = j?.str("enrolled_at"),
        )
    }
}

/** Response from /auth/enrollment-status. */
data class EnrollmentStatus(
    val stakeName: String? = null,
    val stakeId: String? = null,
    val lastSyncedAt: String? = null,
    val memberCount: Int = 0,
    val hasData: Boolean = false,
    val noRole: Boolean = false,
    val credential: CredentialInfo = CredentialInfo(),
) {
    companion object {
        fun from(j: JsonObject) = EnrollmentStatus(
            stakeName = j.str("stake_name"),
            stakeId = j.str("stake_id"),
            lastSyncedAt = j.str("last_synced_at"),
            memberCount = j.int("member_count"),
            hasData = j.bool("has_data"),
            noRole = j.str("status") == "no_role",
            credential = CredentialInfo.from(j.obj("credential")),
        )
    }
}

/**
 * Thin client for the Church-login auth broker (backend/auth_broker). Ported from broker_client.dart
 * — same endpoints, same cold-start retry, same MFA flow. Authed calls carry the signed-in user's
 * Supabase access token (from [auth]). Empty BROKER_URL => `available` is false.
 */
class BrokerClient(private val auth: AuthRepository = AuthRepository()) {

    val available: Boolean get() = AppConfig.brokerAvailable
    private val base: String get() = AppConfig.brokerUrl

    /** Called when a network attempt fails and we're about to retry — UI shows a "waking up…" note. */
    var onStatus: ((String) -> Unit)? = null

    /** N5: wake the free-tier broker early (cheap /health ping) so it's warm by the time the user
     *  submits credentials, hiding the ~60s cold start. Best-effort; errors swallowed. */
    suspend fun warmUp() {
        if (!available) return
        runCatching { Net.getJson("$base/health") }
    }

    // Free hosting (Render) sleeps when idle; the first request after a sleep fails. Retry across
    // ~60s so a cold start resolves itself instead of erroring out. Delays sum to ~63s (matches Dart).
    private val retryDelaysMs = longArrayOf(3000, 6000, 9000, 12000, 15000, 18000)

    private fun requireToken(): String =
        auth.accessToken?.takeIf { it.isNotEmpty() } ?: throw BrokerException("Not signed in.")

    /** POST with cold-start retry; returns the decoded JSON (throws BrokerException on error). */
    private suspend fun postJsonRetry(path: String, body: JsonObject): JsonObject {
        if (!available) throw BrokerException("Church login is not configured (BROKER_URL).")
        var resp: HttpResponse? = null
        var lastErr: Throwable? = null
        for (attempt in 0..retryDelaysMs.size) {
            try {
                resp = Net.postJson("$base$path", body)
                break // got an HTTP response (success or error) — stop retrying
            } catch (e: Throwable) {
                lastErr = e
                if (attempt < retryDelaysMs.size) {
                    onStatus?.invoke("Waking up the sign-in service… this can take up to a minute on first use.")
                    delay(retryDelaysMs[attempt])
                }
            }
        }
        if (resp == null) {
            throw BrokerException(
                "Could not reach the sign-in service after several tries. It may be starting up — " +
                    "please try again in a minute. ($lastErr)",
            )
        }
        val data = Net.bodyObject(resp)
        if (Net.isError(resp)) {
            throw BrokerException(data.str("detail") ?: "Sign-in failed (${resp.status.value}).")
        }
        return data
    }

    private suspend fun post(path: String, body: JsonObject): BrokerResult {
        val data = postJsonRetry(path, body)
        if (data.str("status") == "mfa_required") {
            return BrokerResult(
                loginId = data.str("login_id"),
                factors = data.arr("factors")?.mapNotNull { (it as? JsonObject)?.let(BrokerFactor::from) } ?: emptyList(),
            )
        }
        val session = data.obj("session") ?: JsonObject(emptyMap())
        return BrokerResult(
            email = session.str("email"),
            otp = session.str("otp"),
            name = data.str("identity_name"),
            // N2: only an explicit false blocks; absent/errored enroll → null (don't block).
            authorized = data.obj("enroll")?.boolOrNull("authorized"),
        )
    }

    suspend fun password(username: String, password: String, enroll: Boolean = false): BrokerResult =
        post("/auth/password", buildJsonObject {
            put("username", JsonPrimitive(username)); put("password", JsonPrimitive(password)); put("enroll", JsonPrimitive(enroll))
        })

    suspend fun selectFactor(loginId: String, factorId: String): BrokerResult =
        post("/auth/mfa/select", buildJsonObject {
            put("login_id", JsonPrimitive(loginId)); put("factor_id", JsonPrimitive(factorId))
        })

    suspend fun verifyMfa(loginId: String, code: String, enroll: Boolean = false): BrokerResult =
        post("/auth/mfa/verify", buildJsonObject {
            put("login_id", JsonPrimitive(loginId)); put("code", JsonPrimitive(code)); put("enroll", JsonPrimitive(enroll))
        })

    /** Email-OTP relay (networks that can't reach Supabase directly): broker emails the code. */
    suspend fun emailStart(email: String) {
        postJsonRetry("/auth/email/start", buildJsonObject { put("email", JsonPrimitive(email)) })
    }

    /** Verify the emailed code via the broker → {access_token, refresh_token} for adoptRefreshToken. */
    suspend fun emailVerify(email: String, code: String): JsonObject =
        postJsonRetry("/auth/email/verify", buildJsonObject {
            put("email", JsonPrimitive(email)); put("code", JsonPrimitive(code))
        })

    suspend fun enrollmentStatus(): EnrollmentStatus {
        val resp = Net.getJson("$base/auth/enrollment-status", requireToken())
        val data = Net.bodyObject(resp)
        if (Net.isError(resp)) throw BrokerException(data.str("detail") ?: "Enrollment status failed.")
        return EnrollmentStatus.from(data)
    }

    suspend fun revoke(stakeId: String) {
        authed("POST", "/auth/revoke", buildJsonObject { put("stake_id", JsonPrimitive(stakeId)) })
    }

    /** Provider triggers a sync for their own stake. Returns {coverage_complete, last_synced_at}. */
    suspend fun syncNow(): JsonObject = authed("POST", "/auth/sync-now")

    suspend fun getSchedule(): JsonObject = authed("GET", "/auth/schedule")
    suspend fun setSchedule(hourEt: Int, paused: Boolean): JsonObject =
        authed("POST", "/auth/schedule", buildJsonObject {
            put("hour_et", JsonPrimitive(hourEt)); put("paused", JsonPrimitive(paused))
        })

    suspend fun googleDriveStatus(): JsonObject = authed("GET", "/auth/google/status")
    suspend fun googleDriveStart(): JsonObject = authed("POST", "/auth/google/start")
    suspend fun googleDriveDisconnect(): JsonObject = authed("POST", "/auth/google/disconnect")

    suspend fun contact(subject: String, message: String) {
        authed("POST", "/contact", buildJsonObject {
            put("subject", JsonPrimitive(subject)); put("message", JsonPrimitive(message))
        })
    }

    suspend fun report(): JsonObject = authed("GET", "/report")
    suspend fun emailReport(toEmail: String? = null): JsonObject =
        authed("POST", "/report/email", buildJsonObject {
            if (!toEmail.isNullOrEmpty()) put("to_email", JsonPrimitive(toEmail))
        })

    /** Shared authed request carrying the signed-in user's Supabase token. */
    private suspend fun authed(method: String, path: String, body: JsonObject = JsonObject(emptyMap())): JsonObject {
        if (!available) throw BrokerException("Broker not configured.")
        val token = requireToken()
        val resp = if (method == "GET") Net.getJson("$base$path", token) else Net.postJson("$base$path", body, token)
        val data = Net.bodyObject(resp)
        if (Net.isError(resp)) throw BrokerException(data.str("detail") ?: "Request failed (${resp.status.value}).")
        return data
    }
}
